native_words.asm

```diff
@@ -9929,9 +9887,77 @@ z_swap:         rts
 ; ## "then"  auto  ANS core
         ; """http://forth-standard.org/standard/core/THEN"""
 xt_then:
+                ; we want to write here as the target for the source branch
+                ;     jsr 0branch or an unconditional branch
+                ; orig:
+                ;     lsb msb
+                ; ...
+                ; here:
+                ; if jsr 0branch and here - orig - 2 <= 127 we can use a native branch
+
                 ; Get the address to jump to.
                 jsr xt_here
 
+                ; is it close enough?
+                jsr xt_two_dup
+                jsr xt_swap
+                jsr xt_minus        ; ( C: orig here offset )
+                lda 1,x
+                bne no_opt
+                lda 0,x
+                dea                 ; we want here - orig - 2
+                dea
+                bmi no_opt          ; up to 127 is ok
+
+                sta 0,x             ; remember updated offset
+
+                ; orig also needs to be the target of a zero branch
+                sec
+                lda 4,x
+                sbc #2
+                sta tmp1
+                lda 5,x
+                sbc #0
+                sta tmp1+1          ; tmp1 points to the jsr target preceding orig
+
+                ldy #0
```

Comment:
we could shorten this check by having 0branch users write a signal placeholder address like false=$ffff instead of 0

Reply:
Well, there are no `0branch` users because it's not a word provided in the dictionary anymore (I think it's already gone in the DPANS-94 standard) and it's only here because Tali version 1 had 0branch.  Forth-2012 has a notion of a control-flow stack, but lists it as implementation dependent (and all of the supporting words are optional).  In Tali, the control-flow stack is just the data stack (that is the data stack holds the various addresses while compiling).

Because 0BRANCH and BRANCH are not exposed to the user, you are welcome to do whatever you want to implement/modify their behavior.  I think it's just the flow control words that use them, so `IF`, `CASE` the loops and all of the associated words.

If JMP is not used, then a word could technically be native compiled (all of the branch opcodes are allowed in native compiling).  The issue with JMP is that when you just copy the assembly, the resulting word ends up jumping back into the original word, and not into its copy.  That's what precludes words with IFs and loops from being natively compiled.

From a "pie in the sky" perspective, if we could keep track of whether a word has used JMP or not while being compiled, we could automatically clear the NeverNative flag (it's set by default for all new words) if there are no JMPs.

This looks like a pretty good start, but don't be afraid of completely removing/rewriting BRANCH and 0BRANCH if that makes sense and you can replace it with better functionality.

Reply:
After re-reading your comments, I see you may have meant "routines that use 0branch" when you said "0branch users", although I think my response still works for that.